overlord/devicestate/devicemgr.go

```diff
@@ -679,6 +679,34 @@ func (m *DeviceManager) ensureCloudInitRestricted() error {
 			statusMsg = "failed to transition to done or error state after 5 minutes"
 		}
 
+		// we should always have a model if we are seeded and are not on classic
+		model, err := m.Model()
+		if err != nil {
+			return err
+		}
+
+		// For UC20, we want to always disable cloud-init after it has run on
+		// first boot unless we are in a "real cloud", i.e. not using NoCloud,
+		// or if we installed cloud-init configuration from the gadget
+		if model.Grade() != asserts.ModelGradeUnset {
+			// look for cloud.conf in the installed gadget snap
+			deviceCtx, err := DeviceCtx(m.state, nil, nil)
+			if err != nil {
+				return err
+			}
+
+			gadgetInfo, err := snapstate.GadgetInfo(m.state, deviceCtx)
+			if err != nil {
+				return err
+			}
+
+			if !sysconfig.HasGadgetCloudConf(gadgetInfo.MountDir()) {
+				// no gadget cloud conf so disable cloud-init if it's not
+				// NoCloud
```

Comment:
is the comment right?  we should disable cloud init if it's NoCloud, no?

Reply:
sorry yes the comment is wrong, it should read

```
// no gadget cloud conf, so disable cloud-init if it's NoCloud 
// (it will be restricted for any other cloud to that specific cloud provider datasource)
```